crates/oauth2-types/src/oidc.rs

```diff
@@ -192,44 +287,588 @@ pub struct Metadata {
+    ResponseTypesMissingIdTokenToken,
+
+    /// `authorization_code` is missing from the supported grant types.
+    #[error("missing `authorization_code` grant type")]
+    GrantTypesMissingAuthorizationCode,
+
+    /// `implicit` is missing from the supported grant types.
+    #[error("missing `implicit` grant type")]
+    GrantTypesMissingImplicit,
+
+    /// The given endpoint is missing auth signing algorithm values, but they
+    /// are required because it supports at least one of the `client_secret_jwt`
+    /// or `private_key_jwt` authentication methods.
+    #[error("{0} auth signing algorithm values contain `none`")]
+    MissingAuthSigningAlgValues(&'static str),
+
+    /// `none` is in the given endpoint's signing algorithm values, but is not
+    /// allowed.
+    #[error("{0} signing algorithm values contain `none`")]
+    SigningAlgValuesWithNone(&'static str),
+
+    /// `RS256` is missing from the given endpoint's signing algorithm values.
+    #[error("missing RS256 in {0} signing algorithm values")]
+    SigningAlgValuesMissingRs256(&'static str),
+}
+
+/// Possible extra restrictions on a URL.
+#[derive(Debug, PartialEq, Eq)]
+enum ExtraUrlRestrictions {
+    /// The URL must not contain a fragment.
+    NoFragment,
+
+    /// The URL must not contain a query or a fragment.
+    NoQueryOrFragment,
+}
+
+/// Validate the URL of the field with the given extra restrictions.
+///
+/// The basic restriction is that the URL must use the `https` scheme.
+fn validate_url(
+    field: &'static str,
+    url: &Url,
+    restrictions: Option<ExtraUrlRestrictions>,
+) -> Result<(), ProviderMetadataVerificationError> {
+    if url.scheme() != "https" {
+        return Err(ProviderMetadataVerificationError::UrlNonHttpsScheme(
+            field,
+            url.clone(),
+        ));
+    }
+
+    if let Some(restrictions) = restrictions {
+        if restrictions == ExtraUrlRestrictions::NoQueryOrFragment && url.query().is_some() {
+            return Err(ProviderMetadataVerificationError::UrlWithQuery(
+                field,
+                url.clone(),
+            ));
+        }
+
+        if matches!(
+            restrictions,
+            ExtraUrlRestrictions::NoFragment | ExtraUrlRestrictions::NoQueryOrFragment
+        ) && url.fragment().is_some()
+        {
+            return Err(ProviderMetadataVerificationError::UrlWithFragment(
+                field,
+                url.clone(),
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+/// Validate the algorithm values of the endpoint according to the
+/// authentication methods.
+///
+/// The restrictions are:
+/// - The algorithm values must not contain `none`,
+/// - If the `client_secret_jwt` or `private_key_jwt` authentication methods are
+///   supported, the values must be present.
+fn validate_signing_alg_values_supported(
+    endpoint: &'static str,
+    values: Option<&[JsonWebSignatureAlg]>,
+    methods: Option<&[OAuthClientAuthenticationMethod]>,
```

Comment:
Here your could take `&Option<T>` directly instead of `Option<&T>` to avoid having to call `Option::as_deref()` everywhere